File: src/main/scala/kim/dencka/java/medium/ProductOfArrayExceptSelf.java
package kim.dencka.java.medium;

import java.util.Arrays;

public class ProductOfArrayExceptSelf {
    public static void main(String[] args) {
        int[] nums = {1, 2, 3, 4};
        System.out.println(Arrays.toString(productExceptSelf(nums))); // [24, 12, 8, 6]
        int[] nums2 = {-1, 1, 0, -3, 3};
        System.out.println(Arrays.toString(productExceptSelf(nums2))); // [0, 0, 9, 0, 0]
    }

    public static int[] productExceptSelf(int[] nums) {
        int[] output = new int[nums.length];

        int left = 1;
        for (int i = 0; i < nums.length; i++) {
            output[i] = left;
            left *= nums[i];
        }

        int right = 1;
        for (int i = nums.length - 1; i >= 0; i--) {
            output[i] *= right;
            right *= nums[i];
        }

        return output;
    }
}

File: src/main/scala/kim/dencka/scala/medium/collection/ProductOfArrayExceptSelf.scala
package kim.dencka.scala.medium.collection

object ProductOfArrayExceptSelf extends App {

  // elegant
  def productExceptSelf(nums: Array[Int]): Array[Int] = {
    val _left  = nums.scanLeft(1)(_ * _).init
    val _right = nums.scanRight(1)(_ * _).tail

    _left zip _right map { case (l, r) => l * r }
  }

  // a little bit faster
//  def productExceptSelf(nums: Array[Int]): Array[Int] = {
//    val (zeros, nonZero, product) = nums.foldLeft(0, 1, 1) {
//      case ((z, non, prod), num) if num == 0 => (z + 1, non, prod * num)
//      case ((z, non, prod), num)             => (z, non * num, prod * num)
//    }
//    if (zeros > 1) {
//      Array.fill(nums.length)(0)
//    } else if (zeros == 1) {
//      val result = Array.fill(nums.length)(0)
//      result(nums.indexOf(0)) = nonZero
//      result
//    } else {
//      for (i <- nums.indices) nums(i) = (product * math.pow(nums(i), -1)).toInt
//      nums
//    }
//  }

  println(productExceptSelf(Array(1, 2, 3, 4)).toList)      // output: 24,12,8,6
  println(productExceptSelf(Array(-1, 1, 0, -3, 3)).toList) // output: 0,0,9,0,0

}
